ussomacaroon/client.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+// Pacakge ussomacaroon provides a client that can authenticate with an
+// identity server by discharging macaroons on an Ubuntu SSO server.
+package ussomacaroon
+
+import (
+	"net/url"
+
+	"github.com/juju/httprequest"
+	errgo "gopkg.in/errgo.v1"
+	"gopkg.in/macaroon-bakery.v2-unstable/bakery"
+	"gopkg.in/macaroon-bakery.v2-unstable/httpbakery"
+	"gopkg.in/macaroon.v2-unstable"
+)
+
+// Macaroon returns a macaroon from the identity provider at the given
+// URL, which will have a third-party caveat
+// addressed to an Ubuntu SSO service.
+func Macaroon(url string) (*macaroon.Macaroon, error) {
+	client := &httprequest.Client{}
+	var resp MacaroonResponse
+	if err := client.Get(url, &resp); err != nil {
+		return nil, errgo.Notef(err, "cannot get macaroon")
+	}
+	return resp.Macaroon, nil
+}
+
+// Visitor is an httpbakery.Visitor that will login using a macaroon
+// discharged by an Ubuntu SSO service.
+type Visitor struct {
+	f func(string) (macaroon.Slice, error)
+}
+
+// NewVisitor creates a Vistor which uses a macaroon previously collected
+// with Macaroon and discharged by the requisit Ubuntu SSO service to log
+// in. The discharged macaroon to use will be requested from the given
+// function when required.
+func NewVisitor(f func(url string) (macaroon.Slice, error)) *Visitor {
+	return &Visitor{
+		f: f,
+	}
+}
+
+// VisitWebPage implements httpbakery.Visitor.VisitWebPage by using a
+// macaroon previously discharged by a trusted Ubuntu SSO service as a
+// login token.
+func (v *Visitor) VisitWebPage(client *httpbakery.Client, methodURLs map[string]*url.URL) error {
+	if methodURLs["usso_macaroon"] == nil {
+		return httpbakery.ErrMethodNotSupported
+	}
+	url := methodURLs["usso_macaroon"].String()
+	ms, err := v.f(url)
+	if err != nil {
+		return errgo.Mask(err, errgo.Any)
+	}
+	cl := httprequest.Client{
+		Doer: client,
+	}
+	err = cl.CallURL(url, &LoginRequest{
+		Macaroons: Macaroons{
+			Macaroons: ms,
+		},
+	}, nil)
+	return errgo.Mask(err)
+}
+
+// DischargeUSSOCaveat discharges the given caveat against an Ubuntu SSO
+// service at the given url using the user information in the given
+// email, password and otp.
+func DischargeUSSOCaveat(url string, caveatID []byte, email, password, otp string) (*macaroon.Macaroon, error) {
```

Comment:
Given that this package is ussomacaroon, perhaps this should be just Discharge (or perhaps AcquireDischarge given that's the name used in httpbakery.

Also, I think the signature should probably be:

    func(caveat macaroon.Caveat, email, password, otp string) (*macaroon.Macaroon, error)